util/stats/query_stats.go

```diff
@@ -144,14 +162,45 @@ func (s *SpanTimer) Finish() {
 	}
 }
 
+type Statistics struct {
+	Timers  *QueryTimers
+	Samples *QuerySamples
+}
+
 type QueryTimers struct {
 	*TimerGroup
 }
 
+type QuerySamples struct {
+	// PeakSamples represent the peak number of samples considered while
+	// evaluating a query.
+	PeakSamples int
+	// TotalSamples represents the total number of samples considered
+	// while evaluation a query.
+	TotalSamples int
+}
+
+type Stats struct {
+	TimerStats  QueryTimers
+	SampleStats QuerySamples
+}
+
+// Increment increments the samples count.
+func (qs *QuerySamples) Increment(samples int) {
+	if samples > qs.PeakSamples {
```

Comment:
This isn't right, as this is only how many samples we just processed - not how many are currently held in memory.

Reply:
That is what the total samples is for: to catch and store the total count of samples while querying.

Could you explain in detail what you meant or expected?

Reply:
The PeakSamples should be the highest amount of samples we had in memory at once.

Reply:
Got it. However, I wanted to know that does this issue aims to count the number of samples the query is considering while evaluation (which is similar to the number of points) or does this aims for the number of samples it had to scan (including the ones that were skipped).

Reply:
I'd do it as every input point processed. At the PromQL level, we don't know anything about skipped samples.

Reply:
> I'd do it as every input point processed

Assuming this as when every point is being appended to the vector.

> PeakSamples should be the highest amount of samples we had in memory at once

I hope this means that number of points as whole added to a vector while evaluating a node.

Reply:
> Assuming this as when every point is being appended to the vector.

Before it's passed in I'd say. A rate 1d is more expensive to calculate than a 1h.

> I hope this means the total number of samples fetched from tsdb into memory. Is that correct?

That's part of it, it's also all the outputs that are currently in memory.

Reply:
> That's part of it, it's also all the outputs that are currently in memory.

That means the vales of PeakSamples is monotonicaly increasing?

Reply:
It's a peak value, so it can't go down within a query.

Reply:
Okay got it. Can you differentiate the between of `totalSamples` vs `peakSamples` as per what you think, so that it clears all my doubts. I will makes the changes right away.

Reply:
They're very different things. One is data processed, one is peak memory.

Reply:
If that's the case, for **peak memory**, why not have a function in SeriesIterator as `SamplesCount()` that returns the number of samples across all the series?

Reply:
That's not peak memory, that's input series.

Reply:
> input samples

Aren't the input samples the ones that contribute to memory usage?

Reply:
Yes, but they're generally a pretty small part of it as how we evaluate range vector functions is optimised so we only need to hold the range for one series in memory at a time. It's the intermediate results that are more the issue.